## Flights Data Exploration Challenge

base de datos CSV de vuelos de EU en 2013

In [17]:
import pandas as pd

df_flights = pd.read_csv('flights.csv')#importamos el CSV como df_flights

df_flights.info() # comando para visualizas columnas
 


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271940 entries, 0 to 271939
Data columns (total 20 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Year               271940 non-null  int64  
 1   Month              271940 non-null  int64  
 2   DayofMonth         271940 non-null  int64  
 3   DayOfWeek          271940 non-null  int64  
 4   Carrier            271940 non-null  object 
 5   OriginAirportID    271940 non-null  int64  
 6   OriginAirportName  271940 non-null  object 
 7   OriginCity         271940 non-null  object 
 8   OriginState        271940 non-null  object 
 9   DestAirportID      271940 non-null  int64  
 10  DestAirportName    271940 non-null  object 
 11  DestCity           271940 non-null  object 
 12  DestState          271940 non-null  object 
 13  CRSDepTime         271940 non-null  int64  
 14  DepDelay           271940 non-null  int64  
 15  DepDel15           269179 non-null  float64
 16  CR

## Indice de columnas

- Year: The year of the flight (all records are from 2013)  
- Month: The month of the flight  
- DayofMonth: The day of the month on which the flight departed  
- DayOfWeek: The day of the week on which the flight departed - from 1 (Monday) to 7 (Sunday)  
- Carrier: The two-letter abbreviation for the airline.  
- OriginAirportID: A unique numeric identifier for the departure aiport  
- OriginAirportName: The full name of the departure airport  
- OriginCity: The departure airport city  
- OriginState: The departure airport state  
- DestAirportID: A unique numeric identifier for the destination aiport  
- DestAirportName: The full name of the destination airport  
- DestCity: The destination airport city  
- DestState: The destination airport state  
- CRSDepTime: The scheduled departure time  
- DepDelay: The number of minutes departure was delayed (flight that left ahead of schedule have a negative value)  
- DelDelay15: A binary indicator that departure was delayed by more than 15 minutes (and therefore considered "late")  
- CRSArrTime: The scheduled arrival time  
- ArrDelay: The number of minutes arrival was delayed (flight that arrived ahead of schedule have a negative value)  
- ArrDelay15: A binary indicator that arrival was delayed by more than 15 minutes (and therefore considered "late")  
- Cancelled: A binary indicator that the flight was cancelled  

## objetivo: identificar posibles factores ue afecten la llegada o salida de los vuelos

### - 1 limpieza de datos
    identificar datos nulos o faltantes y reemplazar por algo adecuado
    existen valores atípicos en **ArrDelay** y **DepDelay
    
- ArrDelay: The number of minutes arrival was delayed (flight that arrived ahead of schedule have a negative value)  
- DepDelay: The number of minutes departure was delayed (flight that left ahead of schedule have a negative value)  

In [2]:
#print(df_flights.to_string())
#pd.set_option('display.max_rows', None) #muestra todo 

df_flights

Year  Month  DayofMonth  DayOfWeek Carrier  OriginAirportID  \
0       2013      9          16          1      DL            15304   
1       2013      9          23          1      WN            14122   
2       2013      9           7          6      AS            14747   
3       2013      7          22          1      OO            13930   
4       2013      5          16          4      DL            13931   
...      ...    ...         ...        ...     ...              ...   
271935  2013      9          20          5      VX            13204   
271936  2013      4          19          5      FL            10397   
271937  2013     10          26          6      WN            12191   
271938  2013      5           7          2      HA            13830   
271939  2013      6          11          2      UA            14771   

                               OriginAirportName     OriginCity OriginState  \
0                            Tampa International          Tampa          FL   
1                       Pittsburgh International     Pittsburgh          PA   
2                   Seattle/Tacoma International        Seattle          WA   
3                   Chicago O'Hare International        Chicago          IL   
4                          Norfolk International        Norfolk          VA   
...                                          ...            ...         ...   
271935                     Orlando International        Orlando          FL   
271936  Hartsfield-Jackson Atlanta International        Atlanta          GA   
271937                           William P Hobby        Houston          TX   
271938                           Kahului Airport        Kahului          HI   
271939               San Francisco International  San Francisco          CA   

        DestAirportID                                    DestAirportName  \
0               12478                      John F. Kennedy International   
1               13232                       Chicago Midway International   
2               11278                  Ronald Reagan Washington National   
3               11042                    Cleveland-Hopkins International   
4               10397           Hartsfield-Jackson Atlanta International   
...               ...                                                ...   
271935          12892                          Los Angeles International   
271936          11697            Fort Lauderdale-Hollywood International   
271937          10821  Baltimore/Washington International Thurgood Ma...   
271938          12173                             Honolulu International   
271939          14100                         Philadelphia International   

               DestCity DestState  CRSDepTime  DepDelay  DepDel15  CRSArrTime  \
0              New York        NY        1539         4       0.0        1824   
1               Chicago        IL         710         3       0.0         740   
2            Washington        DC         810        -3       0.0        1614   
3             Cleveland        OH         804        35       1.0        1027   
4               Atlanta        GA         545        -1       0.0         728   
...                 ...       ...         ...       ...       ...         ...   
271935      Los Angeles        CA        1730         1       0.0        1945   
271936  Fort Lauderdale        FL        1020        24       1.0        1218   
271937        Baltimore        MD        1540        30       1.0        1935   
271938         Honolulu        HI         609        -5       0.0         641   
271939     Philadelphia        PA        1338        60       1.0        2200   

        ArrDelay  ArrDel15  Cancelled  
0             13         0          0  
1             22         1          0  
2             -7         0          0  
3             33         1          0  
4             -9         0          0  
...          ...       ...        ...  
271935        18         1          0  


## Datos atípicos
nuestro ejercicio deja en claro que se considera un dato atípico los valores negativos en **ArrDelay**, filtraremos los datos atípicos para visualizarlos

In [3]:
ArrDelay_negativos = df_flights[df_flights['ArrDelay']<0]
ArrDelay_negativos


Year  Month  DayofMonth  DayOfWeek Carrier  OriginAirportID  \
2       2013      9           7          6      AS            14747   
4       2013      5          16          4      DL            13931   
6       2013     10           6          7      WN            13796   
8       2013     10           8          2      AA            13930   
10      2013      6           9          7      OO            12892   
...      ...    ...         ...        ...     ...              ...   
271929  2013      9           9          1      DL            14869   
271931  2013      7           5          5      WN            15304   
271933  2013     10          29          2      US            13930   
271934  2013      4          13          6      AA            11298   
271938  2013      5           7          2      HA            13830   

                         OriginAirportName         OriginCity OriginState  \
2             Seattle/Tacoma International            Seattle          WA   
4                    Norfolk International            Norfolk          VA   
6       Metropolitan Oakland International            Oakland          CA   
8             Chicago O'Hare International            Chicago          IL   
10               Los Angeles International        Los Angeles          CA   
...                                    ...                ...         ...   
271929        Salt Lake City International     Salt Lake City          UT   
271931                 Tampa International              Tampa          FL   
271933        Chicago O'Hare International            Chicago          IL   
271934     Dallas/Fort Worth International  Dallas/Fort Worth          TX   
271938                     Kahului Airport            Kahului          HI   

        DestAirportID                           DestAirportName  \
2               11278         Ronald Reagan Washington National   
4               10397  Hartsfield-Jackson Atlanta International   
6               12191                           William P Hobby   
8               11298           Dallas/Fort Worth International   
10              14679                   San Diego International   
...               ...                                       ...   
271929          10397  Hartsfield-Jackson Atlanta International   
271931          10529                     Bradley International   
271933          14100                Philadelphia International   
271934          13204                     Orlando International   
271938          12173                    Honolulu International   

                 DestCity DestState  CRSDepTime  DepDelay  DepDel15  \
2              Washington        DC         810        -3       0.0   
4                 Atlanta        GA         545        -1       0.0   
6                 Houston        TX         630        -1       0.0   
8       Dallas/Fort Worth        TX        1010         8       0.0   
10              San Diego        CA        1404         3       0.0   
...                   ...       ...         ...       ...       ...   
271929            Atlanta        GA        1655         0       0.0   
271931           Hartford        CT        1100         2       0.0   
271933       Philadelphia        PA        1345        -6       0.0   
271934            Orlando        FL        1335        -7       0.0   
271938           Honolulu        HI         609        -5       0.0   

        CRSArrTime  ArrDelay  ArrDel15  Cancelled  
2             1614        -7         0          0  
4              728        -9         0          0  
6             1210        -3         0          0  
8             1240       -10         0          0  
10            1507        -8         0          0  
...            ...       ...       ...        ...  
271929        2232        -6         0          0  
271931        1350       -10         0          0  
271933        1645       -11         0          0  
271934        1705       -12         0          0  


nuestros dátos atípicos son bastantes:  
152126 filas con datos atípicos en **ArrDelay**   
contra 271940 del total del dataframe, cercano a la mitad


In [4]:
DepDelay_negativos = df_flights[df_flights['DepDelay']<0]
DepDelay_negativos

Year  Month  DayofMonth  DayOfWeek Carrier  OriginAirportID  \
2       2013      9           7          6      AS            14747   
4       2013      5          16          4      DL            13931   
6       2013     10           6          7      WN            13796   
14      2013     10          20          7      AS            14107   
15      2013      4          14          7      AA            12892   
...      ...    ...         ...        ...     ...              ...   
271928  2013      4          14          7      WN            13495   
271932  2013      8          26          1      WN            11292   
271933  2013     10          29          2      US            13930   
271934  2013      4          13          6      AA            11298   
271938  2013      5           7          2      HA            13830   

                                OriginAirportName         OriginCity  \
2                    Seattle/Tacoma International            Seattle   
4                           Norfolk International            Norfolk   
6              Metropolitan Oakland International            Oakland   
14               Phoenix Sky Harbor International            Phoenix   
15                      Los Angeles International        Los Angeles   
...                                           ...                ...   
271928  Louis Armstrong New Orleans International        New Orleans   
271932                       Denver International             Denver   
271933               Chicago O'Hare International            Chicago   
271934            Dallas/Fort Worth International  Dallas/Fort Worth   
271938                            Kahului Airport            Kahului   

       OriginState  DestAirportID                           DestAirportName  \
2               WA          11278         Ronald Reagan Washington National   
4               VA          10397  Hartsfield-Jackson Atlanta International   
6               CA          12191                           William P Hobby   
14              AZ          14747              Seattle/Tacoma International   
15              CA          13830                           Kahului Airport   
...            ...            ...                                       ...   
271928          LA          11259                         Dallas Love Field   
271932          CO          12191                           William P Hobby   
271933          IL          14100                Philadelphia International   
271934          TX          13204                     Orlando International   
271938          HI          12173                    Honolulu International   

            DestCity DestState  CRSDepTime  DepDelay  DepDel15  CRSArrTime  \
2         Washington        DC         810        -3       0.0        1614   
4            Atlanta        GA         545        -1       0.0         728   
6            Houston        TX         630        -1       0.0        1210   
14           Seattle        WA        1330       -11       0.0        1628   
15           Kahului        HI        1730        -7       0.0        2010   
...              ...       ...         ...       ...       ...         ...   
271928        Dallas        TX        1905        -1       0.0        2035   
271932       Houston        TX        1855        -2       0.0        2215   
271933  Philadelphia        PA        1345        -6       0.0        1645   
271934       Orlando        FL        1335        -7       0.0        1705   
271938      Honolulu        HI         609        -5       0.0         641   

        ArrDelay  ArrDel15  Cancelled  
2             -7         0          0  
4             -9         0          0  
6             -3         0          0  
14           -15         0          0  
15            26         1          0  
...          ...       ...        ...  
271928        -6         0          0  
271932         0         0          0  
271933       -11         0          0  
2719

para **DepDelay** también es una cantidad similar

para poder analizar los datos tenemos que trabajar esos datos atípicos 
existen varios métodos para tratar con ellos  

- quitarlos del dataframe
- reemplazarlos con la media
- Winzorisación (remplazo con los valores no atípicos mas cercanos, en este caso 0)

recortar los datos no es conveniente, ya que perdemos casi la mitad del dataframe 

reemplazando con la media conservamos los datos, pero marcaríamos como tardios los vuelos cuando no fueron así

la Winsorización parece mas acertada, eliminamos elementos negativos y conservamos los vuelos que llegaron a tiempo con un valor de 0

antes de tratar los datos contabilizamos si existen valores nulos en cualquier otra columna:

In [5]:
null_values = df_flights.isnull().sum()
null_values

Year                    0
Month                   0
DayofMonth              0
DayOfWeek               0
Carrier                 0
OriginAirportID         0
OriginAirportName       0
OriginCity              0
OriginState             0
DestAirportID           0
DestAirportName         0
DestCity                0
DestState               0
CRSDepTime              0
DepDelay                0
DepDel15             2761
CRSArrTime              0
ArrDelay                0
ArrDel15                0
Cancelled               0
dtype: int64

son muy pocos como para afectar el dataframe, podemos recortarlos

In [6]:
df_not_null_values = df_flights.dropna()
df_not_null_values.isnull().sum()

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
Carrier              0
OriginAirportID      0
OriginAirportName    0
OriginCity           0
OriginState          0
DestAirportID        0
DestAirportName      0
DestCity             0
DestState            0
CRSDepTime           0
DepDelay             0
DepDel15             0
CRSArrTime           0
ArrDelay             0
ArrDel15             0
Cancelled            0
dtype: int64

### ahora podemos aplicar la Winsorisacion 

In [14]:
#df_flights ['ArrDelay'] < 0
df_flights['ArrDelay'] = df_flights['ArrDelay'].where(df_flights['ArrDelay'] >= 0, 0) 
# checamos valores negativos en ArrDelay
ArrDelay_negativos = df_flights[df_flights['ArrDelay']<0]
ArrDelay_negativos
#ninguno
#sustituimos DepDelay
df_flights['DepDelay'] = df_flights['DepDelay'].where(df_flights['DepDelay'] >= 0, 0) 
#checamos DepDelay negativos
DepDelay_negativos = df_flights[df_flights['DepDelay']<0]
DepDelay_negativos

Empty DataFrame
Columns: [Year, Month, DayofMonth, DayOfWeek, Carrier, OriginAirportID, OriginAirportName, OriginCity, OriginState, DestAirportID, DestAirportName, DestCity, DestState, CRSDepTime, DepDelay, DepDel15, CRSArrTime, ArrDelay, ArrDel15, Cancelled]
Index: []

### Según el ejercicio ahora podemos trabajar con los datos 

In [20]:
#limpiar valores nulos antes de 

df_flights = df_flights.dropna()

df_flights.info()
null_values = df_flights.isnull().sum()
null_values

<class 'pandas.core.frame.DataFrame'>
Index: 269179 entries, 0 to 271939
Data columns (total 20 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Year               269179 non-null  int64  
 1   Month              269179 non-null  int64  
 2   DayofMonth         269179 non-null  int64  
 3   DayOfWeek          269179 non-null  int64  
 4   Carrier            269179 non-null  object 
 5   OriginAirportID    269179 non-null  int64  
 6   OriginAirportName  269179 non-null  object 
 7   OriginCity         269179 non-null  object 
 8   OriginState        269179 non-null  object 
 9   DestAirportID      269179 non-null  int64  
 10  DestAirportName    269179 non-null  object 
 11  DestCity           269179 non-null  object 
 12  DestState          269179 non-null  object 
 13  CRSDepTime         269179 non-null  int64  
 14  DepDelay           269179 non-null  int64  
 15  DepDel15           269179 non-null  float64
 16  CRSArrT

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
Carrier              0
OriginAirportID      0
OriginAirportName    0
OriginCity           0
OriginState          0
DestAirportID        0
DestAirportName      0
DestCity             0
DestState            0
CRSDepTime           0
DepDelay             0
DepDel15             0
CRSArrTime           0
ArrDelay             0
ArrDel15             0
Cancelled            0
dtype: int64

## datos listos para trabajar

## ¿que es lo que usualmete causa el retraso de un vuelo?
- condiciones metorológicas
- aerolinea (procesos logísticos, carga de equipaje, problemas mecánicos)
- exceso de tráfico de aviones
- motivos de seguridad
  
estos motivos pueden ser asociados a causas de origen

- condiciones metorológicas (localización del aeropuerto)
- aerolinea (procesos logísticos, carga de equipaje, problemas mecánicos) (aerolinea)
- exceso de tráfico de aviones (localización del aeropuerto)
- motivos de seguridad (irregularidades)

#### los dos factores que mas conviene estudiar son localización del aereopuerto y aerolínea
